### PR TITLE
 Make kubectl exec support "pod/" prefix in the pod name (closes #65670)

### LINKED ISSUE #65670: Make kubectl exec support pod/ prefix in the name provided
/kind feature

I have a curl pod deployed for debugging, and I'd like to be able to exec into it using the output of `kubectl get -oname` like this

```kubectl exec $(kubectl get pods -l app=curl -oname)```

I can't, because `get -oname` returns pod/curl-FOO and kubectl exec does not allow resources with slashes:

```
$ kubectl exec curl-7748d66b69-sdjht /bin/sh
Defaulting container name to part-1.
Use 'kubectl describe pod/curl-7748d66b69-sdjht -n default' to see all of the containers in this pod.
^C

$ kubectl exec pod/curl-7748d66b69-sdjht /bin/sh
error: invalid resource name "pod/curl-7748d66b69-sdjht": [may not contain '/']
```

Client details:

```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T22:29:25Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.4-gke.2", GitCommit:"eb2e43842aaa21d6f0bb65d6adf5a84bbdc62eaf", GitTreeState:"clean", BuildDate:"2018-06-15T21:48:39Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by shal**
**What this PR does / why we need it**:
Fixes #65670

**Release note**:
```release-note
Add support of "pod/" prefix in the pod name to kubectl exec command.
```

### CODE REVIEW

File: pkg/kubectl/cmd/exec.go [hunk 1]
```diff
@@ -159,6 +160,11 @@ func (p *ExecOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []s
 		}
 		p.Command = argsIn
 	} else {
+		if strings.HasPrefix(argsIn[0], "pods/") {
```
Comment (idealhack) [inline pkg/kubectl/cmd/exec.go]: It seems that the [proposing issue](https://github.com/kubernetes/kubernetes/issues/65670) wants the prefix to be `pod/`?

Also, please fix the release note.

**Review by idealhack — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/exec.go (hunk 1 above)
Comment (liggitt) [reply pkg/kubectl/cmd/exec.go]: string.TrimPrefix would be simpler to follow

Comment (idealhack) [reply pkg/kubectl/cmd/exec.go]: +1 for `strings.TrimPrefix`

**Comment by ashanaakh** [comment]
@idealhack I changed prefix to "pod/" as you asked, but I afraid it is wrong. We definitely should use "pods/".

```sh
$> kubectl get pods -oname
```
This command returns:

```
pods/xxxxx-xxxxx-xxxxxxx-86m4s
pods/xxxxx-xxxxx-xxxxxxx-hf57k
```

**Comment by idealhack** [comment]
@ashanaakh I guess the difference is because of different versions. We'd better find the source of this prefix instead of hard-coded it here.

**Comment by idealhack** [comment]
@ashanaakh No, I mean this fix _can_ be done here, we just need to make sure _how_ should it be done.

**Comment by ashanaakh** [comment]
@idealhack Got it, thanks for explanation :smile:

**Comment by idealhack** [comment]
@kubernetes/sig-cli-maintainers

**Comment by mengqiy** [comment]
> Second, why this feature was introduced in #33176 but removed in #33180, my guess is they lost it when resolving conflicts. If this is the case, I'd like we do this by applying the original commit.

Applying the original commit will not work. In the original commit, it was using the `Builder` in kubectl. But now this is no long the case in `exec` code.

IMO using `TrimPrefix` is OK. We need to figure out what is the output on HEAD, `pod/` or `pods/`.
And then make it work with HEAD.

**Comment by idealhack** [comment]
@mengqiy Thanks for clarifying. I didn't mean just cherry-pick it, more like we should include both the implement and the test cases.

**Comment by mengqiy** [comment]
I believe no matter what it prints, it's a client-side behavior. 
The most common use case is `kubectl exec $(kubectl get pods -l app=curl -oname)`, where `kubectl` is the same version. So IMO we only need to ensure it works with HEAD.

I'm fine if you want to do both. It's up to you :)

**Review by idealhack — APPROVED** [review]

File: pkg/kubectl/cmd/exec.go [hunk 2]
```diff
@@ -159,7 +160,8 @@ func (p *ExecOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []s
 		}
 		p.Command = argsIn
 	} else {
-		p.PodName = argsIn[0]
+		// Extract "pod/" from pod name.
+		p.PodName = strings.TrimPrefix(argsIn[0], "pod/")
```
Comment (liggitt) [inline pkg/kubectl/cmd/exec.go]: this should look more like what we do for `kubectl attach`, using resource builder (to support `foo`, `pod/foo`, `pods/foo`, `pods.v1/foo`, etc)

https://github.com/kubernetes/kubernetes/blob/97169b1512db0699bd4f3932ca9660e7659ba063/pkg/kubectl/cmd/attach/attach.go#L222-L245

(that will also mean moving the pod name lookup to the Run method)

**Comment by craigbox** [comment]
@shal is this one close to being ready with @liggitt's feedback?

/remove-lifecycle stale

**Comment by shal** [comment]
@craigbox I'll try to create a new PR shortly according @liggitt comment 😄

**Pull request closed without merging**